docs/config-validation.md

```diff
@@ -0,0 +1,191 @@
+met.
+
+### `nostructlevel`
+
+The same as `structonly`, but it will never run the validations on the nested
+struct, even if one of the `required_` conditions is met. When this tag is
+present, the validator will only validate the value of the struct itself is
+non-nil.
+
+### `min=1`, `gte=1`, `gt=0`
+
+These validate that the value is greater than zero. On its face you might assume
+that these would only validate integers, but when specified on a slice or map it
+will validate that the length of the slice or map is greater than zero.
+
+Note: we encourage the use of `min=1` or `gte=1` over `gt=0` because it's more
+readable.
+
+For instance, the following would be valid config for a slice valued field
+tagged with `required`.
+```json
+{
+  "foo": [],
+}
+```
+
+But, only the following would be valid config for a slice valued field tagged
+with `min=1`.
+```json
+{
+  "foo": ["bar", ...],
+}
+```
+
+### `len`
+
+Same as `eq` (equal to) but can also be used to validate the length of the
+strings.
+
+### `hostname_port`
+
+The
+[docs](https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-HostPort)
+for this tag are scant with detail, but it validates that the value is a valid
+RFC 1123 hostname and port. It is used to validate many of the the
+`ListenAddress` and `DebugAddr` fields of our components.
+
+#### Future Work
+
+This tag is compatible with IPv4 addresses, but not IPv6 addresses. We should
+consider fixing this in our fork of the validator.
+
+### `dive`
+This tag is used to validate the values of a slice or map. For instance, the
+following would be valid config for a slice valued field (`[]string`) tagged
+with `min=1,dive,oneof=bar baz`.
+
+```json
+{
+  "foo": [
+    "bar",
+    "baz",
+  ],
+}
+```
+
+Note that the `dive` tag introduces an order-dependence in writing tags: tags
+that come before `dive` apply to the current field, while tags that come after
+`dive` apply to the current field's child values. In the example above: `min=1`
+applies to the length of the slice (`[]string`), while `oneof=bar baz` applies
+to the value of each string in the slice.
+
+We can also use `dive` to validate the values of a map. For instance, the
+following would be valid config for a map valued field (`map[string]string`)
+tagged with `min=1,dive,oneof=bar baz`.
+
+```json
+{
+  "foo": {
+    "bar": "baz",
+    "baz": "bar"
+  },
+}
+```
+
+`dive` can also be invoked multiple times to validate the values of nested
+slices or maps. For instance, the following would be valid config for a slice of
+slice valued field (`[][]string`) tagged with `min=1,dive,min=2,dive,oneof=bar
+baz`.
+
+```json
+{
+  "foo": [
+    [
+      "bar",
+      "baz",
+    ],
+    [
+      "baz",
+      "bar",
+    ],
+  ],
+}
+```
+
+```go
+foo [][]string `validate:"min=1,dive,gt=1,dive,oneof=bar baz"`
+```
+
+- `min=1` will be applied to the outer slice (`[]`).
+- `gt=1` will be applied to inner slice (`[]string`).
```

Comment:
This example uses both min=1 and gt=1; it should use min=2 instead like on line 136.

Also, I don't think repeating the example here on lines 154-156 is necessary; the reader just saw it on line 136.